Automattic/woocommerce-payments
# Add order items directly to Stripe subscription

Fixes #2893 

#### Changes proposed in this Pull Request

- Add shipping and fee items to a subscription at creation
- Add Stripe coupons to a subscription for WC discounts (This is pending our WCPay accounts being gated into the multiple coupons feature by Stripe)
- Add validation at `invoice.upcoming` webhook event

This changes how subscriptions items are added. These were previously added as invoice items during one of the renewal webhook events. But this PR removes that functionality and instead adds items directly to a subscription when it is created in Stripe.

With this approach, we also have to create coupons for WC discounts since the subscriptions endpoint does not allow for negative item amounts.

Note that going with this approach means we need to add a generic product in Stripe for each of our WC order item types, which we store in `wp_options`.

#### Testing instructions

This PR relies on this WCPay Server PR: 1172-gh-Automattic/woocommerce-payments-server

- Make sure the WCPay Server PR mentioned above is checked out and local server is running
- Make sure you have a subscription product that has been synced to Stripe
- Make sure this product has some shipping cost and or sign-up fee associated with it
- Check out the product and verify the item data in WC and Stripe match.
- Repeat the above for various subscriptions with different settings (free trial, synchronized, etc.)

To test validation, you will need to use Billing Clocks via the dev tools plugin.

- Enable billing clocks on a subscription in WP-Admin (WooCommerce > Subscriptions > your subscription)
- DON'T TRIGGER THE UPCOMING INVOICE YET. Go to your Stripe dashboard and update the quantity and tax rates for all items on the matching subscription: ([#](https://cloudup.com/cBd3CLUjoxf), [#](https://cloudup.com/cqg6JCWd0AX))
- Now trigger the upcoming invoice and confirm the correct values have been set again in Stripe.

-------------------


**Post merge**
